packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts

```diff
@@ -56,59 +56,480 @@ describe("treeApi", () => {
+						type _ = requireAssignableTo<typeof result, number>;
+						assert.equal(result, 43);
+					}
+					{
+						// Returns the special rollback value
+						const result = Tree.runTransaction(root, () => Tree.runTransaction.rollback);
+						type _ = requireAssignableTo<typeof result, symbol>;
+						assert.equal(result, Tree.runTransaction.rollback);
+					}
+					{
+						// Returns either a result or the special rollback value
+						const result = Tree.runTransaction(root, () =>
+							Math.random() >= 0.5 ? Tree.runTransaction.rollback : 43,
+						);
+						if (result === Tree.runTransaction.rollback) {
+							type _ = requireAssignableTo<typeof result, typeof Tree.runTransaction.rollback>;
+						} else {
+							type _ = requireAssignableTo<typeof result, number>;
+						}
+					}
+					{
+						// Returns some symbol or the special rollback value
+						const otherSymbol = Symbol();
+						const result = Tree.runTransaction(root, () =>
+							Math.random() >= 0.5 ? Tree.runTransaction.rollback : otherSymbol,
+						);
+						if (result === Tree.runTransaction.rollback) {
+							type _ = requireAssignableTo<typeof result, typeof Tree.runTransaction.rollback>;
+						} else {
+							type _ = requireAssignableTo<typeof result, typeof otherSymbol>;
+						}
+					}
+				});
+
+				it("emits change events", () => {
+					const { root } = getTestObjectView();
+					let deepEvent = false;
+					let shallowEvent = false;
+					Tree.on(root, "nodeChanged", () => (shallowEvent = true));
+					Tree.on(root, "treeChanged", () => (deepEvent = true));
+					Tree.runTransaction(root, (r) => {
+						r.content = 43;
+					});
+					assert.equal(shallowEvent, true);
+					assert.equal(deepEvent, true);
+				});
+
+				it("emits change events on rollback", () => {
+					const { root } = getTestObjectView();
+					let deepEventCount = 0;
+					let shallowEventCount = 0;
+					Tree.on(root, "nodeChanged", () => (shallowEventCount += 1));
+					Tree.on(root, "treeChanged", () => (deepEventCount += 1));
+					Tree.runTransaction(root, (r) => {
+						r.content = 43;
+						return Tree.runTransaction.rollback;
+					});
+					// One firing of events during the initial change and another during rollback, plus 'treeChanged' fires twice
+					// each time (detach and attach passes).
+					assert.equal(shallowEventCount, 2);
+					assert.equal(deepEventCount, 2);
+				});
+
+				// TODO: When SchematizingSimpleTreeView supports forking, add test coverage to ensure that transactions work properly on forks
+			});
+		});
+
+		describe("view.runTransaction:transaction() return values", () => {
+			it("implicit success, no return value", () => {
 				const view = getTestObjectView();
-				run(view, (root: TestObject) => {
-					assert.equal(root, view.root);
+				const runTransactionResult = view.runTransaction(() => {
+					view.root.content = 43;
 				});
+				assert.equal(view.root.content, 43);
+				assert.equal(runTransactionResult.success, true);
```

Comment:
For all of these tests, I propose doing an asset.deepEqual on the whole `runTransactionResult` object to get more coverage. This will for example ensure that we don't put a bogus `value` in the returned object.

Reply:
done